### ISSUE TITLE
ODBC data type -155 is not supported.

### ISSUE THREAD

**Issue by joshuaduffy** [mkleehammer/pyodbc#134]
Hello,

I'm working with a MSSQL database that uses DATETIMEOFFSET's for date storage. However the type is unsupported, would you consider adding it in? DATETIMEOFFSET is an  "UPPERCASE" type and as I understand quite a common way to store TZ info with a timestamp in SQL Server.

I can work around it with casting, but it would be great to see proper support. Finally, thanks for all your work on this so far as well, really happy with it and it is serving me very well :)

Example:  `2016-08-17 10:08:56.18 +00:00`

I'm using v3.0.10 on Windows with Python 3.5.2 if that helps.

**Comment by mkleehammer** [comment]
Two quick workarounds until I figure out if I can make something official:

1) You can always add an output converter function which will receive the value as a byte array.  You can then convert to the right Python object and return it.
2) I've modified the default behavior to return unknown data types as byte arrays.

**Comment by stevenwinfield** [comment]
If it helps, I just wrote an output converter function for this.

The binary format is this:
`struct.Struct("<6hI2h")`
...which unpacks 20 bytes of data into a tuple of 9 values:
`year, month, day, hours, minutes, seconds, nanoseconds, offset_hours, offset_minutes`
...which can be used to construct a datetime object.

If the timezone offset is negative, both offset_hours and offset_minutes will be negative.

You'll need a tzinfo implementation if you want your datetimes to be timezone aware - FixedOffsetTimezone from psycopg2.tz does the job.
Also remember that datetime's constructor accepts microseconds, not nanoseconds:

`total_offset_minutes = offset_hours * 60 + offset_minutes`
`result = datetime.datetime(year, month, day, hours, minutes, seconds, nanoseconds // 1000, tzinfo=FixedOffsetTimezone(total_offset_minutes))`

**Comment by gordthompson** [comment]
While it's nice to have the option of creating an output converter function when our code uses pyodbc directly, it does still leave cases like [this question on Stack Overflow](http://stackoverflow.com/q/42827492/2144390) where we're at the mercy of what the ORM (or whatever) does with pyodbc on our behalf.

**Comment by akifkamal** [comment]
This is for reading data from the database. What about if we want to store datetime with timezone?

**Comment by v-chojas** [comment]
You can always supply parameters as string and let ODBC driver/database do the conversion.

**Comment by gordthompson** [comment]
Further to comment from @v-chojas I added an example of the required string formatting [here](https://github.com/mkleehammer/pyodbc/wiki/Tips-and-Tricks-by-Database-Platform#datetimeoffset-columns-eg-odbc-sql-type--155-is-not-yet-supported).

**Comment by MichaelE919** [comment]
I ran the following code on a Windows machine with Python 3.6.6:

```python
import pyodbc
cnxn = pyodbc.connect('DSN=devapppbi;Trusted_Connection=yes')
crsr = cnxn.cursor()
query = crsr.execute("select * from dbo.job_master")
results = query.fetchall()
print(results)

crsr.close()
cnxn.close()
```

I ran the same code on a RHEL7 machine with Python 3.6.3:

```python
import pyodbc
cnxn = pyodbc.connect('DSN=devapppbi;Trusted_Connection=yes')
crsr = cnxn.cursor()
query = crsr.execute("select * from dbo.job_master")
results = query.fetchall()
pyodbc.ProgrammingError: ('ODBC SQL type -155 is not yet supported.  column-index=6  type=-155', 'HY106')
```

Both machines are running the same code, same Python version, same connection string, same database and table, and same query. The difference with the RHEL7 machine is it's using unixODBC and Kerberos to connect to Microsoft SQL Server. The results of running this code tells me that ODBC SQL type -155 is supported on the Windows machine but not on the RHEL7 machine. That doesn't make sense- either pyodbc supports it or it doesn't. That has to be some other root cause, but I can't determine it.

Has anyone else seen anything similar to this behavior?

**Comment by v-chojas** [comment]
What version of pyODBC?
Check:
```
import pyodbc
print(pyodbc.version)
```

**Comment by sseveringhaus-newclassrooms** [comment]
My understanding is that exactly what pyodbc supports will depend on exactly what constants are defined in headers like `sql.h`, `sqlext.h`, or `msodbcsql.h` when it's built and installed. That could explain why it works on one system and not another.

In case this is useful to anyone else, here is a short write-up of an investigation into this that I did a while ago while trying to fix this in Django. (I ended up with a janky monkeypatch that I'm not going to recommend using.) I'm not entirely sure all of this is correct at this point, either, and I'd be happily corrected on any of it.

---

SQL type -155 is `SQL_SS_TIMESTAMPOFFSET`, which you can find documented:

https://docs.microsoft.com/en-us/sql/relational-databases/native-client-odbc-date-time/data-type-support-for-odbc-date-and-time-improvements?view=sql-server-2017

Since this is a SQL-Server-specific extension to ODBC, it's defined only in `msodbcsql.h` (e.g. `/opt/microsoft/msodbcsql17/include/msodbcsql.h`), and not in `/usr/include/sql.h` or `sqlext.h` as provided by `unixodbc-dev`.

If you grab the pyodbc source:

https://files.pythonhosted.org/packages/a0/14/d23278740dd64aa47eea396c4caf5a1b464dacaf339dca21b9afbc973e4b/pyodbc-4.0.22.tar.gz

...you can see in `src/pydobc.h` that it's getting its values for those constants like `SQL_SS_TIMESTAMPOFFSET` from `sql.h` and `sqlext.h`, and it uses those (e.g. in `src/getdata.cpp`) to figure out how to process data.

Those in turn get baked into `site-packages/sql_server/pyodbc/introspection.py` when you install the package.

So to fix this "correctly", we would need to patch pyodbc to make it include Microsoft's custom extensions. This seems like an unlikely patch to make it upstream.

So why does this work with older versions of Django/Python/pyodbc? I'm not sure, but I think there might have previously been some kind of automatic or catch-all conversion to another datetime type. I haven't had a chance to go through the commit history and look for that.

**Comment by v-chojas** [comment]
Which ODBC driver is being used is also important, because the server will not advertise the newer types to those older drivers which do not support them:

https://docs.microsoft.com/en-us/sql/relational-databases/native-client-odbc-date-time/enhanced-date-and-time-type-behavior-with-previous-sql-server-versions-odbc

> Client applications that were compiled using a version of SQL Server Native Client prior to SQL Server 2008 see the new date/time types as nvarchar columns.

**Comment by gordthompson** [comment]
Further to comment from @v-chojas:

Situation described by @MichaelE919 is almost certainly that the Windows machine was using the ancient "SQL Server" driver (shipped with Windows) while the Linux machine was using a newer "ODBC Driver xx for SQL Server". The former returns the value as a string, while the latter returns a type -155 value which requires an [output converter function](https://github.com/mkleehammer/pyodbc/issues/134#issuecomment-281739794) to interpret the results. The issue is also documented in the wiki [here](https://github.com/mkleehammer/pyodbc/wiki/Tips-and-Tricks-by-Database-Platform#datetimeoffset-columns-eg-odbc-sql-type--155-is-not-yet-supported).

**Comment by kristang** [comment]
Is this an issue that can arise when writing to SQL? I am experiencing and issue with writing datetimeoffsets from a pandas dataframe to an MSSQL table, and I need a fairly specific combination of versions of pyodbc (4.0.22)  together with one of SQL Server Native Client 11.0 or 10.0.

If I use the latest (4.0.25 as of now) the datetimeoffsets are stripped of their offsets and inserted into MSSLQ without them

**Comment by gordthompson** [comment]
@kristang - Yes. Currently when pyodbc writes to a `datetimeoffset` column the offset will be dropped if the parameter value is a timezone-aware `datetime` object:

```python
crsr.execute("CREATE TABLE ##tmp (id INT PRIMARY KEY, dto DATETIMEOFFSET)")
sql = "INSERT INTO ##tmp (id, dto) VALUES (?, ?)"
dto = datetime(2018, 8, 2, 0, 28, 12, tzinfo=timezone(timedelta(hours=-6)))
params = (1, dto)
crsr.execute(sql, params)
print(crsr.execute("SELECT id, CAST(dto AS VARCHAR) FROM ##tmp").fetchall())
# [(1, '2018-08-02 00:28:12.0000000 +0')]
```

The workaround is to supply the parameter value as a string, as described [here](https://github.com/mkleehammer/pyodbc/wiki/Tips-and-Tricks-by-Database-Platform#datetimeoffset-columns-eg-odbc-sql-type--155-is-not-yet-supported).

**Comment by kristang** [comment]
Forgive my ignorance, but is this an issue internally to pyodbc, or the specific ODBC drivers I am using? 

For example, with pyodbc 4.0.22, I can write datetimeoffsets to SQL when using SQL Server Native Client 11.0 or 10.0. However, when reading with the same configuration (from MSSQL to a `pandas` dataframe) i encounter the type -155 error. 

Conversely, changing to using "SQL Server" will allow me to read correctly into a dataframe, but will strip the offset when trying to write.

**Comment by gordthompson** [comment]
@kristang 

re: "I can write datetimeoffsets to SQL when using SQL Server Native Client 11.0 or 10.0."

How are you performing the writes? My example above was run using "ODBC Driver 11 for SQL Server" and the offset was stripped. Are you using pandas+SQLAlchemy to push a dataframe to SQL Server?

re: the ancient "SQL Server" ODBC driver

That driver was written around the time of SQL Server 2000, so `datetimeoffset` columns didn't exist yet. As is often the case, older ODBC drivers that don't have native support for newer data types will do their best to handle them as strings. So if you query a `datetimeoffset` column with the older "SQL Server" driver you'll get a string (similar to the CAST in my example above) while a newer ODBC driver will return the column as ODBC type -155 [SQL_SS_TIMESTAMPOFFSET]. pyodbc does not (yet) include explicit support for SQL_SS_TIMESTAMPOFFSET, possibly because it is a SQL_Server-specific extension to ODBC.

**Comment by kristang** [comment]
Yes. This is done through `pandas` with the supplied function: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_sql.html

**Comment by James-DBA-Anderson** [comment]
For anyone else looking to fix this issue when using an ORM (I'm using SQLAlchemy), switching the driver from 'SQL+Server+Native+Client+11.0' to the ancient 'SQL+Server' seems to have worked. 

I'm not sure of any side effects that this may cause but I'm able to read DATETIMEOFFSET values and write them with the offset intact.

**Comment by zzzeek** [comment]
SQLAlchemy here .

How we doing on this?   I'm getting hit on this now.     I'm using the Microsoft drivers on linux.

**Comment by vdurante** [comment]
Is there any way of developing some kind of input converter? I am having issues on INSERT and would like to have a central solution that will convert the `datetime` to a string with timezone instead of doing this on every single insert command in my code. Thank you.

**Comment by gordthompson** [comment]
@vdurante  - I've been working on an [enhancement to SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy/issues/4983) to add `datetimeoffset` support to the `mssql+pyodbc` dialect. I'm hoping that it will be included in the next release. See [this comment](https://github.com/sqlalchemy/sqlalchemy/issues/4983#issuecomment-554327454) for (IMO) the underlying reason why support for `datetimeoffset` is not baked into pyodbc itself.

**Comment by bwesen** [comment]
The horror, while searching for an error ("-155 not supported") in a seemingly completely trivial piece of code (reading ONE line using pyodbc from an MSSQL server) and finding an Open ticket 5 years old :/

So I just installed pyodbc on an Ubuntu 20.04LTS after installing the latest MS SQL Odbc driver from their own repos, and also had to install the unixodbc-dev package to get pip install pyodbc to work. 

I used the instructions here: https://docs.microsoft.com/sv-se/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server?view=sql-server-ver15

The server is "Microsoft SQL Azure (RTM) - 12.0.2000.8" and the driver is "ODBC Driver 17 for SQL Server".

The database uses datetimeoffset(7) and this field crashes cursor.fetchone(), unexpectedly I should add :)

I read somewhere above that you could use an output type converter, but how do I apply that if fetchone() doesn't return anything I can convert? Or do you add it into pyodbc some way before executing the query?

**Comment by gordthompson** [comment]
@bwesen - You register your output converter function using `Connection.add_output_converter()`. Details here:

https://github.com/mkleehammer/pyodbc/wiki/Using-an-Output-Converter-function

**Comment by haider101** [comment]
> @bwesen - You register your output converter function using `Connection.add_output_converter()`. Details here:
> 
> https://github.com/mkleehammer/pyodbc/wiki/Using-an-Output-Converter-function

worked like a charm 👍

**Comment by jdeus** [comment]
Any update?

**Comment by gordthompson** [comment]
> Any update?

No, and a built-in fix for this issue in pyodbc is increasingly unlikely now that the future of SQL Server access from Python belongs to

https://pypi.org/project/mssql-python/

**Comment by Nandu825** [comment]
is this issue fixed? i tried using pyodbc, aioodbc (which is essentially an async wrapper on pyodbc) but both fail at this - date time offset. mssql python smoothly supports it and it has no issues. but the problem is that it doesnt have native async support. any workaround for that? as i need async connection.

**Comment by gordthompson** [comment]
@Nandu825 - SQLAlchemy can handle `datetimeoffset` values under async:

```python
import asyncio

from sqlalchemy.ext.asyncio import create_async_engine


async def async_main():
    engine = create_async_engine(
        "mssql+aioodbc://scott:tiger^5HHH@msodbcsql18_199",
    )
    async with engine.begin() as conn:
        result = await conn.exec_driver_sql(
            "select cast('2026-03-03 08:04:02 -07:00' as datetimeoffset) as dto"
        )
        print(repr(result.scalar()))
        """
datetime.datetime(2026, 3, 3, 8, 4, 2, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=61200)))
        """

asyncio.run(async_main())
```
